### ISSUE TITLE
GaussProductExperiment generates nan values for distributions with support of size 2

### ISSUE THREAD

**Issue by cottoncini** [openturns/openturns#1756]
Hi everyone,
a GaussProductExperiment built from a ComposedDistribution in which at least one of the marginals' support size is 2 generates samples with nan values:

    import openturns as ot
    d1 = ot.UserDefined(ot.Sample([[10], [20]]))
    d2 = ot.Bernoulli()
    d3 = ot.UserDefined(ot.Sample([[10], [100], [1000]]))
    dc = ot.ComposedDistribution([d1, d2, d3])
    print(ot.GaussProductExperiment(dc, [2, 2, 3]).generate())

         [ X0   X1   X2   ]
     0 : [  nan  nan   10 ]
     1 : [ -nan  nan   10 ]
     2 : [  nan -nan   10 ]
     3 : [ -nan -nan   10 ]
     4 : [  nan  nan  100 ]
     5 : [ -nan  nan  100 ]
     6 : [  nan -nan  100 ]
     7 : [ -nan -nan  100 ]
     8 : [  nan  nan 1000 ]
     9 : [ -nan  nan 1000 ]
    10 : [  nan -nan 1000 ]
    11 : [ -nan -nan 1000 ]

Please refer to the thread at https://gitter.im/openturns/community?at=6017e4249238c531ad186b54
Thank you!

**Comment by jschueller** [comment]
Replacing isIntegral by isDriscrete in AdaptiveStieltjesAlgorithm is necessary but it does not solve the bug.

Both the krawtchouk & the stieljes give inf for n=1 for the Bernouilli:
```
distribution = ot.Bernoulli()
algo0 = ot.KrawtchoukFactory(1, 0.5)
algo1 = ot.AdaptiveStieltjesAlgorithm(distribution)
iMax = distribution.getSupport().getSize()
for i in range(iMax):
    ott.assert_almost_equal(algo1.getRecurrenceCoefficients(i), algo0.getRecurrenceCoefficients(i))
```

**Comment by jschueller** [comment]
i think we should ensure "n_>n" in KrawtchoukFactory instead "n_>=n"

**Comment by regislebrun** [comment]
@jschueller You are right, there is also a bug in KrawtchoukFactory: the reccurence coefficients of order n allow to compute P_{n+1}! So in the case of a Bernoulli distribution, i.e. when n=1, you can only ask for n=0. But you can ask the StandardDistributionPolynomialFactory  class for the polynomial of degree 0 (always constant and equal to 1, no recurrence coefficients) and degree 1 (recurrence coefficients with n=0).
I will check the other issues you mention.

**Comment by jschueller** [comment]
@regislebrun I added the Stieljes isIntegral modification & Krawtchouk fix in #1766 
in my branch the snippet from cottoncini throws as expected, but its not over

- AdaptiveStieltjesAlgorithm returns an inf for this code:
```
print(ot.AdaptiveStieltjesAlgorithm(ot.Bernoulli()).getRecurrenceCoefficients(2))
```
it comes from a zero division at https://github.com/openturns/openturns/blob/master/lib/src/Uncertainty/Algorithm/OrthogonalBasis/AdaptiveStieltjesAlgorithm.cxx#L104

- As I said earlier, recurrence coefficients from NegativeBinomial/MeixnerFactory, Poisson/Charlier do not match:
```
distribution = ot.NegativeBinomial()
algo0 = ot.MeixnerFactory(1, 0.5)
algo1 = ot.AdaptiveStieltjesAlgorithm(distribution)
iMax = distribution.getSupport().getSize()
iMax = 10 # FIXME
for i in range(iMax):
    print(i, algo1.getRecurrenceCoefficients(i), algo0.getRecurrenceCoefficients(i))

distribution = ot.Poisson()
algo0 = ot.CharlierFactory(1.0)
algo1 = ot.AdaptiveStieltjesAlgorithm(distribution)
iMax = distribution.getSupport().getSize()
#iMax = 1 # FIXME
for i in range(iMax):
    print(i, algo1.getRecurrenceCoefficients(i), algo0.getRecurrenceCoefficients(i))
```
